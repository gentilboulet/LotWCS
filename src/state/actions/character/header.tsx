import { createAction } from "typesafe-actions";
import * as dataArchetypes from "../../../data/archetypes";
import * as dataRanks from "../../../data/ranks";
import { IPerk } from "../../../perks";
import { historyMetaCreator } from "../meta";

export const setName = createAction(
  "header/SET_NAME",
  (name: string) => {
    return { name };
  },
  historyMetaCreator,
)();

export const setConcept = createAction(
  "header/SET_CONCEPT",
  (concept: string) => {
    return { concept };
  },
  historyMetaCreator,
)();

export const setArchetype = createAction(
  "header/SET_ARCHETYPE",
  (archetype: dataArchetypes.TArchetype) => {
    dataArchetypes.validateArchetype(archetype);
    return { archetype };
  },
  historyMetaCreator,
)();

export const setRank = createAction(
  "header/SET_RANK",
  (rank: dataRanks.TRank) => {
    dataRanks.validateRank(rank);
    const dr = dataRanks.getRank(rank) as dataRanks.IDataRank;
    return { rank, perks: dr.perks as IPerk[] };
  },
  historyMetaCreator,
)();
