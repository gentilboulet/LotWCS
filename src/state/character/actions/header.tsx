import { createAction } from "typesafe-actions";

import * as dataArchetypes from "../../../data/archetypes";
import * as dataRanks from "../../../data/ranks";

export const setName = createAction(
  "header/SET_NAME",
  action => (name: string) => {
    return action({ name });
  },
);

export const setConcept = createAction(
  "header/SET_CONCEPT",
  action => (concept: string) => {
    return action({ concept });
  },
);

export const setArchetype = createAction(
  "header/SET_ARCHETYPE",
  action => (archetype: dataArchetypes.TArchetype) => {
    dataArchetypes.validateArchetype(archetype);
    return action({ archetype });
  },
);

export const setRank = createAction(
  "header/SET_RANK",
  action => (rank: dataRanks.TRank) => {
    dataRanks.validateRank(rank);
    return action({ rank });
  },
);
