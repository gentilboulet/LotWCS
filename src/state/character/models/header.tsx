import * as data from "../../../data/ranks";
import { ICharacterState } from "./type";

export function setRank(state: ICharacterState, rank: data.TRank): void {
  const dataRank = data.getRank(rank) as data.IDataRank;
  state.rank = dataRank.value;
}
