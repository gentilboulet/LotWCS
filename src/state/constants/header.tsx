export const HEADER_SET_NAME = 'HEADER_SET_NAME';
export type HEADER_SET_NAME = typeof HEADER_SET_NAME;

export const HEADER_SET_CONCEPT = 'HEADER_SET_CONCEPT';
export type HEADER_SET_CONCEPT = typeof HEADER_SET_CONCEPT;

export const HEADER_SET_ARCHETYPE = 'HEADER_SET_ARCHETYPE';
export type HEADER_SET_ARCHETYPE = typeof HEADER_SET_ARCHETYPE;

export const HEADER_SET_RANK = 'HEADER_SET_RANK';
export type HEADER_SET_RANK = typeof HEADER_SET_RANK;
