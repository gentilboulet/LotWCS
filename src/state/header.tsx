import { ranks } from 'data/ranks';
import { IStoreState } from 'state/type';

// Sub Reducers
import { applyBonuses, IBonus, isBonus } from 'state/bonuses';
import { IDiscount, isDiscount, pushToDiscounts } from 'state/discounts';

export function setRank(state: IStoreState, rank: string): void {
  const idx = ranks.findIndex((data) => (data.key === rank));
  state.rank = { name: rank, value: ranks[idx].value };
  applyBonuses(state, ranks[idx].perks.filter(p => isBonus(p)) as IBonus[]);
  pushToDiscounts(state, ranks[idx].perks.filter(p => isDiscount(p)) as IDiscount[]);
}
