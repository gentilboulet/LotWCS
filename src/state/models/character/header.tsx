import * as data from "../../../data/ranks";
import { ICharacterState } from "./index";

export function setRank(state: ICharacterState, rank: data.TRank): void {
  const dataRank = data.getRank(rank) as data.IDataRank;
  state.rank = dataRank.value;
}
