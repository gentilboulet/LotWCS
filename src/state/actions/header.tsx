import * as constants from 'state/constants/header';

import * as dataArchetypes from 'data/archetypes';
import * as dataRanks from 'data/ranks';

export interface IHeaderSetName {
  type: constants.HEADER_SET_NAME;
  name: string;
}

export interface IHeaderSetConcept {
  type: constants.HEADER_SET_CONCEPT;
  concept: string;
}

export interface IHeaderSetArchetype {
  type: constants.HEADER_SET_ARCHETYPE;
  archetype: string;
}

export interface IHeaderSetRank {
  type: constants.HEADER_SET_RANK;
  rank: string;
}

export type IHeaderAction = IHeaderSetName | IHeaderSetConcept | IHeaderSetArchetype | IHeaderSetRank;

export function setName(name: string): IHeaderSetName {
  return { type: constants.HEADER_SET_NAME, name };
}

export function setConcept(concept: string): IHeaderSetConcept {
  return { type: constants.HEADER_SET_CONCEPT, concept };
}

export function setArchetype(archetype: string): IHeaderSetArchetype {
  dataArchetypes.validateArchetype(archetype);
  return { archetype, type: constants.HEADER_SET_ARCHETYPE };
}

export function setRank(rank: string): IHeaderSetRank {
  dataRanks.validateRank(rank);
  return { rank, type: constants.HEADER_SET_RANK };
}
