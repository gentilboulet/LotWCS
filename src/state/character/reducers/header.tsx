import { produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../actions/header";
import { setRank } from "../models/header";
import { ICharacterState } from "../models/type";

export function headerReducer(
  state: ICharacterState,
  action: ActionType<typeof actions>,
): ICharacterState {
  return produce(state, draft => {
    switch (action.type) {
      case getType(actions.setName):
        draft.name = action.payload.name;
        break;
      case getType(actions.setConcept):
        draft.concept = action.payload.concept;
        break;
      case getType(actions.setRank):
        setRank(draft, action.payload.rank);
        break;
      case getType(actions.setArchetype):
        draft.archetype = action.payload.archetype;
        break;
      default:
    }
  });
}
