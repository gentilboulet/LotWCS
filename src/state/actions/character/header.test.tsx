import { TArchetype } from "../../../data/archetypes";
import { TRank } from "../../../data/ranks";
import * as actions from "./header";

describe("Testing header action creators", () => {
  test("should create a set name action", () => {
    const name = "New Name";
    const a = actions.setName(name);
    expect(a.payload.name).toBe(name);
    expect(a).toMatchSnapshot();
  });

  test("should create a set concept action", () => {
    const concept = "New Concept";
    const a = actions.setConcept(concept);
    expect(a.payload.concept).toBe(concept);
    expect(a).toMatchSnapshot();
  });

  test("should create a set archetype action", () => {
    const archetype = "doctor";
    const a = actions.setArchetype(archetype);
    expect(a.payload.archetype).toBe(archetype);
    expect(a).toMatchSnapshot();
  });

  test("should not create an invalid set archetype action", () => {
    expect(() =>
      actions.setArchetype(("pro golfer" as unknown) as TArchetype),
    ).toThrowError();
  });

  test("should create a set rank action", () => {
    const rank = 2;
    const a = actions.setRank(rank);
    expect(a.payload.rank).toBe(rank);
    expect(a).toMatchSnapshot();
  });

  test("should not create an invalid set rank action", () => {
    expect(() =>
      actions.setRank(("ultimate rank" as unknown) as TRank),
    ).toThrowError();
  });
});
