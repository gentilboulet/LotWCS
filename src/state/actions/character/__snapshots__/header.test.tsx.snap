// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing header action creators should create a set archetype action 1`] = `
Object {
  "meta": Object {
    "inHistory": true,
  },
  "payload": Object {
    "archetype": "doctor",
  },
  "type": "header/SET_ARCHETYPE",
}
`;

exports[`Testing header action creators should create a set concept action 1`] = `
Object {
  "meta": Object {
    "inHistory": true,
  },
  "payload": Object {
    "concept": "New Concept",
  },
  "type": "header/SET_CONCEPT",
}
`;

exports[`Testing header action creators should create a set name action 1`] = `
Object {
  "meta": Object {
    "inHistory": true,
  },
  "payload": Object {
    "name": "New Name",
  },
  "type": "header/SET_NAME",
}
`;

exports[`Testing header action creators should create a set rank action 1`] = `
Object {
  "meta": Object {
    "inHistory": true,
  },
  "payload": Object {
    "perks": Array [
      Object {
        "type": "BONUS_DESTINY",
        "value": 20,
      },
      Object {
        "skills": Array [
          "Awareness",
          "Confidence",
          "Crafting",
          "Finesse",
          "Hardiness",
          "Inspire",
          "Learning",
          "Medicine",
          "Might",
          "Perform",
          "Politics",
          "Ride",
          "Stealth",
          "Survival",
          "Tactics",
          "Wu Wei",
        ],
        "type": "DISCOUNT_SKILL",
        "value": 20,
      },
      Object {
        "chi": "general",
        "type": "BONUS_CHI",
        "value": 10,
      },
      Object {
        "kfType": "KUNGFU_EXTERNAL",
        "type": "DISCOUNT_KUNGFU_STYLE",
        "uids": Array [],
        "value": 10,
      },
      Object {
        "kfType": "KUNGFU_INTERNAL",
        "type": "DISCOUNT_KUNGFU_STYLE",
        "uids": Array [],
        "value": 10,
      },
      Object {
        "type": "DISCOUNT_VIRTUE",
        "value": 75,
        "virtues": Array [
          Object {
            "name": "Honor",
            "type": "VIRTUE_CHIVALROUS",
          },
          Object {
            "name": "Benevolence",
            "type": "VIRTUE_CHIVALROUS",
          },
          Object {
            "name": "Righteousness",
            "type": "VIRTUE_CHIVALROUS",
          },
          Object {
            "name": "Loyalty",
            "type": "VIRTUE_CHIVALROUS",
          },
          Object {
            "name": "Force",
            "type": "VIRTUE_CHIVALROUS",
          },
          Object {
            "name": "Revenge",
            "type": "VIRTUE_SELFISH",
          },
          Object {
            "name": "Individualism",
            "type": "VIRTUE_SELFISH",
          },
          Object {
            "name": "Obsession",
            "type": "VIRTUE_SELFISH",
          },
          Object {
            "name": "Ruthlessness",
            "type": "VIRTUE_SELFISH",
          },
          Object {
            "name": "Ferocity",
            "type": "VIRTUE_SELFISH",
          },
        ],
      },
    ],
    "rank": 2,
  },
  "type": "header/SET_RANK",
}
`;
